src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs

```diff
@@ -229,20 +231,103 @@ private static void WriteChunksTo(SourceText sourceText, ObjectWriter writer, in
+
+        using var chunkReader = new CharArrayChunkTextReader(chunks, totalLength);
+        var result = SourceText.From(chunkReader, totalLength, encoding, checksumAlgorithm);
+        Contract.ThrowIfTrue(totalLength != chunkReader.Position);
+
+        return result;
+
+        static ImmutableArray<char[]> CreateChunks(int totalLength)
+        {
+            var numberOfChunks = 1 + (totalLength / CharArrayLength);
+            var buffer = new FixedSizeArrayBuilder<char[]>(numberOfChunks);
+            for (var i = 0; i < numberOfChunks; i++)
+                buffer.Add(s_charArrayPool.Allocate());
+
+            return buffer.MoveToImmutable();
+        }
+    }
+
+    private static int GetIndexFromPosition(int position) => position / CharArrayLength;
+    private static int GetColumnFromPosition(int position) => position % CharArrayLength;
+
+    private sealed class CharArrayChunkTextWriter(
+        int totalLength, ImmutableArray<char[]> chunks, Encoding encoding, CancellationToken cancellationToken) : TextWriter
+    {
+        private readonly int _totalLength = totalLength;
+        private readonly ImmutableArray<char[]> _chunks = chunks;
+        private readonly CancellationToken _cancellationToken = cancellationToken;
+
+        /// <summary>
+        /// Public so that caller can assert that writing out the text actually wrote out the full text of the node.
+        /// </summary>
+        public int Position;
+
+        public override Encoding Encoding { get; } = encoding;
+
+        public override void Write(string? value)
+        {
+            Contract.ThrowIfNull(value);
+
+            var valueSpan = value.AsSpan();
+            while (valueSpan.Length > 0)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                var chunk = _chunks[GetIndexFromPosition(Position)];
+                Contract.ThrowIfTrue(chunk.Length != CharArrayLength);
+
+                var chunkIndex = GetColumnFromPosition(Position);
+                Contract.ThrowIfTrue(chunkIndex >= CharArrayLength);
+
+                var count = Math.Min(valueSpan.Length, CharArrayLength - chunkIndex);
+                valueSpan[..count].CopyTo(chunk.AsSpan().Slice(chunkIndex, count));
+
+                Position += count;
+                valueSpan = valueSpan[count..];
+            }
+
+            Contract.ThrowIfTrue(Position > _totalLength);
+        }
+    }
+
+    private sealed class CharArrayChunkTextReader(ImmutableArray<char[]> chunks, int length) : TextReaderWithLength(length)
     {
-        private readonly ImmutableArray<char[]> _chunks;
-        private readonly int _chunkSize;
-        private bool _disposed;
+        private readonly ImmutableArray<char[]> _chunks = chunks;
+        private bool _disposed = false;
 
-        private int _position;
+        /// <summary>
+        /// Public so that the caller can assert that the new SourceText read all the way to the end of this successfully.
```

Comment:
>/// Public so that the caller can assert that the new SourceText read all the way to the end of this successfully. [](http://example.com/codeflow?start=8&length=114)

nit: does it need to be writable for the callers too?

Reply:
no.  will change to a prop with private setter.

Reply:
Done.